src/app-layer-ssl.c

```diff
@@ -250,54 +255,416 @@ static void SSLSetTxDetectFlags(void *vtx, uint8_t dir, uint64_t flags)
     }
 }
 
-static int TLSDecodeHandshakeHello(SSLState *ssl_state, uint8_t *input,
-                                   uint32_t input_len)
+/**
+ * \inline
+ * \brief Check if value is GREASE.
+ *
+ * http://tools.ietf.org/html/draft-davidben-tls-grease-00
+ *
+ * \param value Value to check.
+ *
+ * \retval 1 if is GREASE.
+ * \retval 0 if not is GREASE.
+ */
+static inline int TLSDecodeValueIsGREASE(const uint16_t value)
+{
+    switch (value)
+    {
+        case 0x0a0a:
+        case 0x1a1a:
+        case 0x2a2a:
+        case 0x3a3a:
+        case 0x4a4a:
+        case 0x5a5a:
+        case 0x6a6a:
+        case 0x7a7a:
+        case 0x8a8a:
+        case 0x9a9a:
+        case 0xaaaa:
+        case 0xbaba:
+        case 0xcaca:
+        case 0xdada:
+        case 0xeaea:
+        case 0xfafa:
+            return 1;
+        default:
+            return 0;
+    }
+}
+
+static inline int TLSDecodeHSHelloVersion(SSLState *ssl_state, uint8_t *input,
+                                          const uint32_t input_len)
 {
     uint8_t *initial_input = input;
 
-    /* only parse the message if it is complete */
-    if (input_len < ssl_state->curr_connp->message_length || input_len < 40)
-        return 0;
+    if (!(HAS_SPACE(SSLV3_CLIENT_HELLO_VERSION_LEN))) {
+        SCLogDebug("TLS handshake invalid length");
+        SSLSetEvent(ssl_state,
+                    TLS_DECODER_EVENT_HANDSHAKE_INVALID_LENGTH);
+        return -1;
+    }
+
+    if ((ssl_state->current_flags & SSL_AL_FLAG_STATE_CLIENT_HELLO) &&
+            ssl_config.enable_ja3) {
+        uint16_t version = *input << 8 | *(input + 1);
+
+        ssl_state->ja3_str = Ja3BufferInit();
+        if (ssl_state->ja3_str == NULL)
+            return -1;
+
+        int rc = Ja3BufferAddValue(ssl_state->ja3_str, version);
+        if (rc != 0)
+            return -1;
+    }
 
-    /* skip version */
     input += SSLV3_CLIENT_HELLO_VERSION_LEN;
 
-    /* skip random */
+    return (input - initial_input);
+}
+
+static inline int TLSDecodeHSHelloRandom(SSLState *ssl_state, uint8_t *input,
+                                         const uint32_t input_len)
+{
+    uint8_t *initial_input = input;
```

Comment:
this can probably be a const to make clear it shouldn't ever change: actually a 'const uint8 * const', as neither the ptr itself as the mem being pointed to should change. (probably applies to many funcs below too)